### ISSUE TITLE
ArcadePhysics GroupVsGroup collisions skip certain objects.

### ISSUE THREAD

**Issue by BenjaminDRichards** [photonstorm/phaser#5011]
## Version

* Phaser Version: 3.22.0

* Operating system: MacOS

* Browser:

## Description

Expected behavior: Tilemap in one Group will collide with Sprite in another Group, if the Groups are set to collide.

Actual behaviour: No collision occurs. Sprite falls through the floor or moves through a wall.

Additional notes:
- Arcade Physics is used in all cases.
- The Group is the Physics version.
- When adding a Tilemap to a Physics Group, it gains a Body. Remove the Body - it does not reflect per-tile collisions.

Cause:
In `Phaser.Physics.Arcade.World.prototype.collideGroupVsGroup`, it steps through the children of group1 and runs `collideSpriteVsGroup`on each, against group2. Thus, when a Tilemap is a child of group1 (but not group2), it is treated as a Sprite.

However, `collideSpriteVsGroup` will return early if it cannot find `body` on the "Sprite". We have removed the Body from the Tilemap to get accurate collisions, so it fails this check, and no collision is run.

## Example Test Code

It's actually easiest to see the problem in the source code:

```js
collideGroupVsGroup: function (group1, group2, collideCallback, processCallback, callbackContext, overlapOnly)
    {
        if (group1.length === 0 || group2.length === 0)
        {
            return;
        }

        var children = group1.getChildren();

        for (var i = 0; i < children.length; i++)
        {
            this.collideSpriteVsGroup(children[i], group2, collideCallback, processCallback, callbackContext, overlapOnly);
        }
    }
```

`collideSpriteVsGroup` has the following code:

```js
if (group.length === 0 || !bodyA || !bodyA.enable)
{
    return;
}
```

Tilemaps do not have bodies, so they won't pass the check, and no collision will be evaluated.

## Additional Information

I resolved the issue in my code by modifying `collideGroupVsGroup`. It uses `collideHandler` instead of `collideSpriteVsGroup`. This correctly evaluates collisions.

I think it could be more efficient, however. `collideHandler` will evaluate the type of group2 every time, but we know it's a Group by this point. This means an unnecessary check of group2 type for every object in group1, which can add up.

A more efficient fix would be a version of `collideHandler` where the second object is known to be a Group.

**Comment by samme** [comment]
What is the `collide()` or `collider()` call exactly?

**Comment by BenjaminDRichards** [comment]
@samme I just create a Collider, `scene.physics.add.collider( group1, group2 )`. If group1 has a Tilemap in it, it won't collide with anything in group2. If group1 has a Sprite in it, it WILL collide with anything in group2.

**Comment by samme** [comment]
I tried collisions with `TilemapLayer` in `Group` and they all seemed to work: <https://codepen.io/samme/pen/MWwePJQ>.

**Comment by BenjaminDRichards** [comment]
Interesting. My discovery conditions were pretty deep in a complex process, and I used `this.physics.add.group` instead of `this.add.group`. If you switch that in the codepen, everything goes crazy. The tilemap falls off screen, because the physics group gives it a body. So I add `this.physics.world.disableBody( layer.body ); delete layer.body;`. At that point, the sprites both fall through the tilemap. I can re-establish collisions by adding `this.physics.add.collider( layer, group2 );`, but the inverse (`this.physics.add.collider( player1, group1 );`) does not work.

If I insert the following function (indicating where a change was made), the group1->group2 collision works, but the group2->group1 collision does not.

```
function fixTilemapGroupCollisions() {
  const World = Phaser.Physics.Arcade.World.prototype;
  if ( !World.__collideGroupVsGroup ) {
      World.__collideGroupVsGroup = World.collideGroupVsGroup;
      World.collideGroupVsGroup = function ( group1, group2, collideCallback, processCallback, callbackContext, overlapOnly ) {

          if (group1.length === 0 || group2.length === 0)
          {
              return;
          }

          var children = group1.getChildren();

          for (var i = 0; i < children.length; i++)
          {
              // Changed from `collideSpriteVsGroup`.
              this.collideHandler(children[i], group2, collideCallback, processCallback, callbackContext, overlapOnly);
          }
      };
  }
}
fixTilemapGroupCollisions()
```

The reason that 1->2 works with this variant code is that `collideHandler` takes the Tilemap from group1 and feeds it to `collideGroupVsTilemap`. But 2->1 doesn't work, because when `collideHandler` gets a Group as parameter `object1`, it only runs a collision if `object2.body` exists. And we have to remove the body to get the Tilemap behaving properly inside the physics group.

In other words, the collision behaves differently depending on whether Tilemap was the first or second participant. That is inconsistent and causes problems. In fact, any time a bodiless Tilemap is the second participant in a Group collision, it looks like it should fail to collide.

Hmm, I thought my `fixTilemapGroupCollisions` method would work, but it only fixes 1->2 collisions. I think I have to fix the other direction too.

**Comment by BenjaminDRichards** [comment]
In fact, `collideSpriteVsGroup` just does `body` checks, so if the Group has a Tilemap, it will be skipped.

I have an idea about swapping the inputs to `collideHandler`. That might make the whole process a bit more robust. Hold on...

**Comment by BenjaminDRichards** [comment]
I made a simplified collision handler. It adds a new function, `collideTypes`, which strictly orders collision participants, breaks down Groups, and boils everything down to either SpriteVsSprite or SpriteVsTilemap right there. This appears to resolve all my issues.

Drawbacks: It invalidates a lot of the methods on World, because it only uses two of them. Furthermore, it can't apply the `useTree` functionality in World, because the tree function assumes that everything in a Group has a Body, and that's not true when dealing with tilemaps. But it's a great deal more solid, and will always run the same interaction no matter which object comes first. (This may be another drawback! It may reorder the objects involved in a collision. If that order is somehow used elsewhere, this code will break that assumption.)

Anyway, here's the modified codepen, with the Phaser hacks at the top:

```js
Phaser.Physics.Arcade.World.prototype.collideHandler = function (object1, object2, collideCallback, processCallback, callbackContext, overlapOnly)
{
    //  Collide Group with Self
    //  Only collide valid objects
    if (object2 === undefined && object1.isParent)
    {
        return this.collideTypes(object1, object1, collideCallback, processCallback, callbackContext, overlapOnly);
    }

    //  If neither of the objects are set then bail out
    if (!object1 || !object2)
    {
        return false;
    }

    return this.collideTypes( object1, object2, collideCallback, processCallback, callbackContext, overlapOnly );
};

Phaser.Physics.Arcade.World.prototype.collideTypes = function(object1, object2, collideCallback, processCallback, callbackContext, overlapOnly)
            {
                // Enforce comparison order.
                // Groups are always first, then tilemaps, then bodies.
                if ( (object1.body && !object2.body) || (!object1.isParent && object2.isParent) ) {
                    var swap = object2;
                    object2 = object1;
                    object1 = swap;
                }
            
                // BODIES
                if (object1.body)
                {
                    if (object2.body) {
                        return this.collideSpriteVsSprite(object1, object2, collideCallback, processCallback, callbackContext, overlapOnly);
                    }
                    // After sorting, object2 is either a body or invalid.
                }
            
                // TILEMAP LAYERS
                else if (object1.isTilemap)
                {
                    if (object2.body)
                    {
                        return this.collideSpriteVsTilemapLayer(object2, object1, collideCallback, processCallback, callbackContext, overlapOnly);
                    }
                    // After sorting, object2 is either a tilemap, body, or invalid.
                    // Tilemaps do not interact, so we only need to evaluate bodies.
                }
            
                // GROUPS
                else if (object1.isParent)
                {
                    var children = object1.getChildren();
                    if (children.length === 0)
                    {
                        return false;
                    }
                    // Unfold Groups for evaluation.
                    var didCollide = false;
                    for (var i = 0; i < children.length; i++ )
                    {
                        if ( this.collideTypes( children[i], object2, collideCallback, processCallback, callbackContext, overlapOnly ) )
                        {
                            didCollide = true;
                        }
                    }
                    return didCollide;
                }
            };

console.clear();

document.getElementById('version')
  .textContent = 'Phaser v' + Phaser.VERSION;

var config = {
  width: 800,
  height: 600,
  physics: {
    default: 'arcade',
    arcade: {
      gravity: { y: 600 }
    }
  },
  scene: {
    preload: preload,
    create: create
  },
  loader: {
    baseURL: 'https://labs.phaser.io',
    crossOrigin: 'anonymous'
  },
};

var game = new Phaser.Game(config);

function preload() {
  this.load.image('tiles', 'assets/tilemaps/tiles/gridtiles.png');
  this.load.tilemapTiledJSON('map', 'assets/tilemaps/maps/simple-map.json');
  this.load.image('player', 'assets/sprites/phaser-dude.png');
}

function create() {
  var map = this.make.tilemap({ key: 'map', tileWidth: 32, tileHeight: 32 });
  var tileset = map.addTilesetImage('tiles');
  var layer = map.createStaticLayer('Level1', tileset);

  map.setCollision([20, 48]);

  var player1 = this.physics.add.image(256, 320, 'player');
  var player2 = this.physics.add.image(320, 320, 'player').setFlipY(true);

  var group1 = this.physics.add.group();
  var group2 = this.physics.add.group();

  group1.add(layer);
  this.physics.world.disableBody( layer.body );
  delete layer.body;

  group2.add(player1);
  group2.add(player2);

  // this.physics.add.collider(group1, player1);
  // this.physics.add.collider(player2, group1);
  this.physics.add.collider(group1, group2);
  // this.physics.add.collider(group2, group1);
}
```

**Comment by BenjaminDRichards** [comment]
It may not be enormously performant, mind you. I'm getting some performance hitches when comparing the fix to unfixed code in a heavy project.

**Comment by samme** [comment]
If the problem occurs only when TilemapLayers are given physics bodies, I don't think it's a bug.

**Comment by BenjaminDRichards** [comment]
Correction: it's not about TilemapLayer bodies. That's just a side effect of the rather subtle situation where it occurs, and I'm removing the bodies anyway.

It's about Group collision functions assuming that the members of a Group are all Sprites (or body-having entities). But the Group is specified to take any kind of GameObject, which includes TilemapLayers, which have different intrinsic collision properties to Sprites. This means that, depending on the order in which groups are compared, tilemap collisions may or may not occur.

(If a tilemap has a Body, the Body is an axis-aligned bounding box equal to the total area of the tilemap. This is not how tilemaps should work, so I disable the body.)

It's possible that Groups should not accept TilemapLayers, but as it stands, they are valid members, and they _sometimes_ work properly. Again, it depends on order.

**Comment by samme** [comment]
For me the order doesn't seem to matter: [a tilemap layer in a group always collides](https://codepen.io/samme/pen/MWwePJQ) and [a tilemap layer in a physics group never collides](https://codepen.io/samme/pen/mdJrEox) (after the body is removed).

It's true that physics groups are more restrictive than ordinary groups. Every member must have a physics body of the same type as the group — so no tilemap layers.

**Comment by BenjaminDRichards** [comment]
Having finally had the time to refactor our physics code...

Things behave fine with the normal group. The physics group was not necessary for physics. In fact, the physics group was detrimental to tilemap physics. I don't know why we went with physics groups in the first place; it's possible that we assumed they were necessary to add collisions between groups. This was untrue; the sole purpose of a physics group is to give its members bodies. The docs actually imply this, but apparently not firmly enough to get it through my skull.

I might suggest making the instructions on physics groups a bit clearer about their sharp focus, and advise that they're not necessary for running physics into a group. Thus others would be spared our confustication.
